### PR TITLE
VM Refactoring

### PR DESCRIPTION
**Pull request by phobos2077**
1. Show script name on description_p_proc (for debugging).
2. VM type system refactoring: make one class for all types, all stacks will contain values directly. No pointers and dynamic_casts. Closes https://github.com/falltergeist/falltergeist/issues/293.
3. Moved push/pop shorthand methods from VM to VMStack class.
4. Move push/pop shorthand methods to VMStack class
5. Refactoring of arithmetic, logical and comparison operators. Added a few missing handlers. For https://github.com/falltergeist/falltergeist/issues/294
6. Made operators work properly with all types, including strings and objects.
7. Merged all comparison operator handlers into one class to avoid complete code duplication.
8. Added missing opcode handlers for bwxor and bwnot.
9. Made obj_lock, obj_unlock handlers not crashy.
10. Added missing "float constant" opcode handler.
11. Added properties sourceObject, targetObject, fixedParam and actionUsed properties to VM class.
12. Properly implemented handlers for source_obj, target_obj, obj_being_used_with, fixed_param, action_being_used.
13. Introduced _warning() and _error() protected methods to OpcodeHandler class - replacing all code of manually throwing exceptions and printing warnings. Now all errors related to script execution will throw VMErrorException. For https://github.com/falltergeist/falltergeist/issues/314
14. Minor refactoring of several opcode handlers.
15. Added locked & closed properties to GameContainerItemObject + fixed obj_lock, obj_unlock handlers to use them.
16. Added sounds when opening/closing doors.

### CODE REVIEW

**Comment by phobos2077** [comment]
This issue seems to be fixed: https://github.com/falltergeist/falltergeist/issues/277

File: src/Game/ContainerItemObject.cpp [hunk 1]
```diff
@@ -52,5 +52,25 @@ void GameContainerItemObject::use_p_proc()
     Game::getInstance()->pushState(state);
 }
 
+void GameContainerItemObject::setLocked(bool _locked)
```
Comment (alexeevdv) [inline src/Game/ContainerItemObject.cpp]: why not "bool locked"? why we need '_' here? :)

File: src/VM/Handlers/OpcodeComparisonHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Falltergeist.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// C++ standard includes
+
+// Falltergeist includes
+#include "../../Logger.h"
+#include "../../VM/Handlers/OpcodeComparisonHandler.h"
```
Comment (alexeevdv) [inline src/VM/Handlers/OpcodeComparisonHandler.cpp]: There is no such word as "Comparison" :)

File: src/Game/ContainerItemObject.cpp (hunk 1 above)
Comment (phobos2077) [inline src/Game/ContainerItemObject.cpp]: That's just how netbeans is doing it :)

File: src/VM/Handlers/OpcodeComparisonHandler.cpp (hunk 2 above)
Comment (phobos2077) [inline src/VM/Handlers/OpcodeComparisonHandler.cpp]: http://prntscr.com/6aslg0

**Comment by alexeevdv** [comment]
I've got this error during compilation process:
`
/home/alexeevdv/falltergeist/src/VM/Handlers/Opcode8106Handler.cpp: In member function ‘virtual void Falltergeist::Opcode8106Handler::_run()’:
/home/alexeevdv/falltergeist/src/VM/Handlers/Opcode8106Handler.cpp:58:60: error: call of overloaded ‘push(std::vector<Falltergeist::Game::GameItemObject*>::size_type)’ is ambiguous
         _vm->dataStack()->push(critter->inventory()->size());
                                                            ^
`

**Pull request merged**